nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1115 @@
+        mat_xml = self.matrix._to_xml_element()
+        if mat_xml is not None:
+            cifti.append(mat_xml)
+        return cifti
+
+    @classmethod
+    def from_header(klass, header=None):
+        if header is None:
+            return klass()
+        if type(header) == klass:
+            return header.copy()
+        raise ValueError('header is not a Cifti2Header')
+
+    @classmethod
+    def may_contain_header(klass, binaryblock):
+        from .parse_cifti2_fast import _Cifti2AsNiftiHeader
+        return _Cifti2AsNiftiHeader.may_contain_header(binaryblock)
+
+
+class Cifti2Image(FileBasedImage):
+    """ Class for single file CIfTI2 format image
+    """
+    header_class = Cifti2Header
+    valid_exts = Nifti2Image.valid_exts
+    files_types = Nifti2Image.files_types
+    makeable = False
+    rw = True
+
+    def __init__(self, data=None, header=None, nifti_header=None):
+        ''' Initialize image
+
+        The image is a combination of (array, affine matrix, header, nifti_header),
+        with optional metadata in `extra`, and filename / file-like objects
+        contained in the `file_map` mapping.
+
+        Parameters
+        ----------
+        dataobj : object
+           Object containg image data.  It should be some object that retuns an
+           array from ``np.asanyarray``.  It should have a ``shape`` attribute
+           or property
+        affine : None or (4,4) array-like
+           homogenous affine giving relationship between voxel coordinates and
+           world coordinates.  Affine can also be None.  In this case,
+           ``obj.affine`` also returns None, and the affine as written to disk
+           will depend on the file format.
+        header : Cifti2Header object
+        nifti_header : None or mapping or nifti2 header instance, optional
+           metadata for this image format
+        '''
+        self._header = header or Cifti2Header()
+        self.data = data
+        self.extra = nifti_header
+
+    def get_data(self):
+        return self.data
+
+    @property
+    def shape(self):
+        return self.data.shape
+
+    @classmethod
+    def from_file_map(klass, file_map):
+        """ Load a Cifti2 image from a file_map
+
+        Parameters
+        file_map : file_map
+
+        Returns
+        -------
+        img : Cifti2Image
+            Returns a Cifti2Image
+         """
+        from .parse_cifti2_fast import _Cifti2AsNiftiImage, Cifti2Extension
+        nifti_img = _Cifti2AsNiftiImage.from_file_map(file_map)
+
+        # Get cifti2 header
+        for item in nifti_img.header.extensions:
+            if isinstance(item, Cifti2Extension):
+                cifti_header = item.get_content()
+                break
+        else:
+            raise ValueError('Nifti2 header does not contain a CIFTI2 '
+                             'extension')
+
+        # Construct cifti image
+        cifti_img = Cifti2Image(data=nifti_img.get_data()[0, 0, 0, 0],
```

Comment:
It would be good to optimize this, because this will read all the data from disk.  `nifti_img.dataobj`, if just read from disk, will be an `ArrayProxy` that points to the data on disk.  By allowing `ArrayProxy` to change shape, by dropping 1-length dimensions, we could keep this as an `ArrayProxy`, and then avoid reading all the data before creating the `Cifti2Image`.